lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -270,6 +314,26 @@ void ESItoRTLPass::runOnOperation() {
     signalPassFailure();
 }
 
+void ESItoRTLPass::removeWraping() {
+  auto top = getOperation();
+
+  // Set up a conversion and give it a set of laws.
+  ConversionTarget target(getContext());
+  target.addLegalDialect<circt::rtl::RTLDialect>();
+  target.addDynamicallyLegalOp<WrapValidReady, UnwrapValidReady>(
+      doesNotHaveEsiPort);
+  target.addIllegalDialect<ESIDialect>();
+
+  // Add all the conversion patterns.
+  ESIRTLBuilder esiBuilder(top);
+  OwningRewritePatternList patterns;
+  patterns.insert<RemoveWrapUnwrap>(&getContext());
```

Comment:
Would it make sense for this to be a canonicalization pattern or folder? I'm not exactly sure if that's how you want to use it, but DialectConversion might be able to just fold these away instead of needing to run a separate conversion.

Reply:
Honestly, I'm still figuring out the dialect conversion framework. How would I specify either?

Reply:
Instead of wasting your time, I'll do something nobody does: read the documentation!

Reply:
This is what I was looking at https://mlir.llvm.org/docs/Canonicalization/#canonicalizing-with-fold

Reply:
Read through all of the applicable documentation and I think that treating this as a fold isn't workable, but it could be a canonicalization pattern.

Reply:
I converted it to a canonicalizer, but hit a crash in MLIR. Trying to move to latest before I write up a bug or fix, but that update packs a punch: #342

Reply:
Cool, I think that might make sense too, depending on how you want to use it. Incidentally, I was debugging the dialect conversion framework myself last night, and I think I see what the docs mean about "fold is used within the dialect conversion infrastructure as a legalization mechanism"... before the framework applies any patterns, it tries to fold the operation:

```
//===-------------------------------------------===//                                                                                                                                                              
Legalizing operation : 'rtl.module'(0x1bd6120) {                                                                                                                                                                   
  * Fold {                                                                                                                                                                                                         
  } -> FAILURE : unable to fold                                                                                                                                                                                    
                                                                                                                                                                                                                   
  * Pattern : 'rtl.module -> ()' {
[...]                                                                                                                                                                                 
```

If, for example, there was a folder that matched for `rtl.module` above, it would have been used by dialect conversion first, before trying out the pattern I wrote.

I guess my understanding is this: folders are for patterns you want applied automatically by dialect conversion, while canonicalizers are for patterns you want applied whenever a canonicalization pass is run.

I'm curious what the crash is...

I'll try to help with the LLVM update... it was something I've been putting off.

Reply:
Yeah, if this is "always profitable" and a peephole optimization on the SSA graph, a canonicalization pattern is probably the right way to go.

Explicit passes are good when you have more heavy analysis required or when the transformation is not always profitable.